cryptoki/src/session/object_management.rs

```diff
@@ -3,66 +3,249 @@
+            // swallow the return value, as we can't do anything about it
+            let _ = unsafe { f(self.session.handle()) };
         }
+    }
+}
+
+impl Session {
+    /// Iterate over session objects matching a template.
+    ///
+    /// # Arguments
+    ///
+    /// * `template` - The template to match objects against
+    ///
+    /// # Returns
+    ///
+    /// This function will return a [`Result<ObjectHandleIterator>`] that can be used to iterate over the objects
+    /// matching the template. Note that the cache size is managed internally and set to a default value (10)
+    ///
+    /// # See also
+    ///
+    /// * [`ObjectHandleIterator`] for more information on how to use the iterator
+    /// * [`Session::iter_objects_with_cache_size`] for a way to specify the cache size
+    #[inline(always)]
+    pub fn iter_objects(&self, template: &[Attribute]) -> Result<ObjectHandleIterator> {
+        self.iter_objects_with_cache_size(template, NonZeroUsize::new(MAX_OBJECT_COUNT).unwrap())
+    }
 
-        let objects = objects.into_iter().map(ObjectHandle::new).collect();
+    /// Iterate over session objects matching a template, with cache size
+    ///
+    /// # Arguments
+    ///
+    /// * `template` - The template to match objects against
+    /// * `cache_size` - The number of objects to cache (type is [`NonZeroUsize`])
+    ///
+    /// # Returns
+    ///
+    /// This function will return a [`Result<ObjectHandleIterator>`] that can be used to iterate over the objects
+    /// matching the template. The cache size corresponds to the size of the array provided to `C_FindObjects()`.
+    ///
+    /// # See also
+    ///
+    /// * [`ObjectHandleIterator`] for more information on how to use the iterator
+    /// * [`Session::iter_objects`] for a simpler way to iterate over objects
+    pub fn iter_objects_with_cache_size(
+        &self,
+        template: &[Attribute],
+        cache_size: NonZeroUsize,
+    ) -> Result<ObjectHandleIterator> {
+        let template: Vec<CK_ATTRIBUTE> = template.iter().map(Into::into).collect();
+        ObjectHandleIterator::new(self, template, cache_size)
+    }
 
-        Ok(objects)
+    /// Search for session objects matching a template
+    ///
+    /// # Arguments
+    ///
+    /// * `template` - The template to match objects against
+    ///
+    /// # Returns
+    ///
+    /// Upon success, a vector of [`ObjectHandle`] wrapped in a Result.
+    /// Upon failure, the first error encountered.
+    ///
+    /// It is a convenience function that will call [`Session::iter_objects`] and collect the results.
+    ///
+    /// # See also
+    ///
+    /// * [`Session::iter_objects`] for a way to specify the cache size
+    #[inline(always)]
+    pub fn find_objects(&self, template: &[Attribute]) -> Result<Vec<ObjectHandle>> {
+        self.iter_objects(template)?
+            .collect::<Result<Vec<ObjectHandle>>>()
```

Comment:
I wonder if this would suffice:

```suggestion
            .collect()
```

(since it's in a return position and the type of the result and this is identical)